### PR TITLE
[WIP] Full Code Review

### PR DESCRIPTION
**Pull request by iamnp**
API-centered code review of the Depot project

### CODE REVIEW

**Review by iamnp — COMMENTED** [review]
# TL;DR

- Overall, the code is good, with several improvements of varying severity suggested, mostly aimed at improving the quality of API usage
- Most of the decisions made by the project authors are reasonable and in line with the project's objectives
- Two critical issues have been identified that require particular attention:
    - `OmnibusPlan` has fields with deferred completion that may be ambiguous to use: [link](https://github.com/lidofinance/depot/pull/24#discussion_r1817743493)
    - The omnibus test file structure is very spread out and contains many boilerplate code, giving a huge field for errors: [link](https://github.com/lidofinance/depot/pull/24#discussion_r1817771517)

> [!NOTE]
> **Severity ratings:**
> **⚠️ - CRITICAL**
> **🔴 - HIGH**
> **🟡 - MEDIUM**
> **👍 - GOOD IDEA**

# API Design

## **Requirements**

APIs should be simple, concise, and easy to use. API should remain minimalist to reduce the amount of code the user must write. The framework should encapsulate complex logic to simplify API usage. Functions and objects should have descriptive names that clearly reflect their purpose and usage.

## **Review Points**

Check that API functions and parameters are self-explanatory and require minimal documentation. Look for any extraneous methods or properties that could be removed or encapsulated within the framework. Look for areas where logic could be encapsulated, such as transaction handling or error management.

## **Findings**

The project appears to be logically structured, with modular directories organised by functionality, including configurations, contracts, keystores, voting and trace handling, which helps with navigation and finding specific functionality. The naming conventions are generally well aligned with the expected functionality (e.g. providers, omnibuses, traces). **Some directories, such as `omnibuses', contain broad functionality (e.g. tools, checks, blueprints) which might benefit from more granular sub-modules to clarify purpose at a glance.** Moreover, enhancing function and parameter names, as suggested, will make the API more intuitive, leading to easier adoption and fewer errors during implementation. See code-related findings tagged with **[API Design].**

# Error Prevention

## **Requirements**

APIs should be bug-proof with no room for ambiguity, making it difficult for developers to misunderstand them. Design APIs to prevent misuse using intuitive naming conventions.

## **Review Points**

Ensure that all API functions handle errors gracefully, with clear messages or error types for developers. Evaluate whether all exposed functions are logically grouped and that any complex functionality is encapsulated to simplify usage.

## **Findings**

The codebase uses custom error classes, error handling and type-safe definitions. **However, improvements could be made by extending the use of exhaustive checks for unsupported conditions.** This would make the APIs more resilient to misuse and minimize potential edge cases that could lead to unhandled errors. See code-related findings tagged with **[Error Prevention]**.

# Code Quality

## **Requirements**

Ensure robust TypeScript type definitions across the codebase to catch potential type errors at compile time. Each API function should include concise documentation outlining usage, parameters, and any edge cases.

## **Review Points**

Check that all functions and variables are appropriately typed and that type inference is used wherever possible. Ensure function and module-level documentation is present and clear.

## **Findings**

Overall, typing and type inference are used effectively within the project, with little room for improvement. The design seems to focus on encapsulating complex interactions. The presence of README files and examples of configuration files indicates some attention to usability. **However, many files lack module and/or function level documentation, it may not be extremely useful to document everything, but focus on API documentation should be a priority.** See code-related findings tagged with **[Code Quality]**.

# Testing

## **Requirements**

The code should be rigorously tested, focusing on API behaviour under various edge cases. Where applicable, tests should mock Ethereum transactions to simulate and validate API behaviour.

## **Review Points**

Ensure that tests cover a wide range of scenarios, including edge cases, expected behaviour and potential misuse. Verify that tests include mock transactions to ensure that the framework behaves predictably under simulated conditions.

## **Findings**

It looks like mock data and utilities, including stubs and simulated transaction environments, are being used effectively. **It is important to control test coverage and increase it from 60% to 80%+.**



**As a general suggestion, clear guidelines and conventions should be introduced for writing both omnibuses and, more importantly, tests.**

**Ability to run omnibus items one-by-one will decrease development time**

File: omnibuses/_example_omnibus.spec.ts [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    let agentLDOBalanceBefore: any;
+    let balancesBefore: BigNumberish[];
+    let nodeOperatorsCountBefore: any;
+
+    before(async () => {
+      agentLDOBalanceBefore = await contracts.ldo.balanceOf(contracts.agent.address);
+      balancesBefore = await Promise.all(tokenTransfers.map(({ recipient }) => contracts.ldo.balanceOf(recipient)));
+      nodeOperatorsCountBefore = await contracts.curatedStakingModule.getNodeOperatorsCount();
+
+      // Start and enact omnibus. Keep receipt to check events later.
```
[the remark this answers is not in the document]
Comment (iamnp) [reply omnibuses/_example_omnibus.spec.ts:115]: Proposed structure of a test file (pseudocode):
Option 1 (less code):
```
const addr = "0x0..."

func beforeVote(beforeState) {
    // Run checks for the whole omnibus before enacting and save artifacts to state object

    // item1
    beforeState.balance1 = checkAddressAmount(addr)
    assert beforeState.balance1 == 10 ETH

    // item2
    beforeState.balance1 = checkAddressAmount(addr)
    assert beforeState.balance1 == 10 ETH
}

func afterVote(beforeState) {
    // Run checks for the whole omnibus after enacting

    // item1
    balance1 = checkAddressAmount(addr)
    assert beforeState.balance1 == balance1

    // item2
    balance1 = checkAddressAmount(addr)
    assert beforeState.balance1 == balance1
}
```

Option 2 (more structured):
```
// comments for item1
item1 = {
    addr = "0x0...",
    beforeVote: (beforeState) -> {
        // Run checks for the item1 before enacting and save artifacts to state object
        beforeState.balance1 = checkAddressAmount(addr)
        assert beforeState.balance1 == 10 ETH
    },
    afterVote: (beforeState) -> {
        // Run checks for the item1 after enacting
        balance1 = checkAddressAmount(addr)
        assert beforeState.balance1 == balance1
    }
}

// comments for item2
item2 = {
    addr = "0x0...",
    beforeVote: (beforeState) -> {
        // Run checks for the item2 before enacting and save artifacts to state object
        beforeState.balance1 = checkAddressAmount(addr)
        assert beforeState.balance1 == 10 ETH
    },
    afterVote: (beforeState) -> {
        // Run checks for the item2 after enacting
        balance1 = checkAddressAmount(addr)
        assert beforeState.balance1 == balance1
    }
}

export [item1, item2]
```

What is two consider:
- Do we have to store data between before- and after- test functions (e.g. use beforeState object)?

File: omnibuses/_example_omnibus.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        topUp: "0x85d703B2A4BaD713b596c647badac9A1e95bB03d",
+        addRecipient: "0x1dCFc37719A99d73a0ce25CeEcbeFbF39938cF2C",
+        removeRecipient: "0x00BB68a12180a8f7E20D8422ba9F81c07A19A79E",
+      },
+      registry: "0xAa47c268e6b2D4ac7d7f7Ffb28A39484f5212c2A",
+    }),
+    blueprints.tokens.transferLDO({
+      title: "Transfer 110,000 LDO to Argo Technology Consulting Ltd. (ATC) multisig",
+      to: "0x9B1cebF7616f2BC73b47D226f90b01a7c9F86956", // Argo Technology Consulting Ltd. (ATC) multisig
+      amount: 110_000n * 10n ** 18n,
```
[the remark this answers is not in the document]
Comment (iamnp) [reply omnibuses/_example_omnibus.ts:31]: Need an utility function that converts "110k LDO" to wei

File: src/omnibuses/blueprints/easy-track.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+interface RemovePaymentEvmScriptFactoryInput {
+  title: string;
+  factory: Address;
+}
+
+function removeEvmScriptFactory(
+  contracts: Contracts<Lido>,
+  { factory, title }: RemovePaymentEvmScriptFactoryInput,
+): OmnibusItem {
+  const { easyTrack, callsScript, voting } = contracts;
```
[the remark this answers is not in the document]
Comment (iamnp) [reply src/omnibuses/blueprints/easy-track.ts:29]: Check `factory, title` against common mistakes (0 address, etc.)
